nipype/interfaces/utility/wrappers.py

```diff
@@ -25,8 +25,91 @@
+        from collections import namedtuple
+        from inspect import getargspec
+        args, varargs, keywords, defaults = getargspec(func)
+        ret_t = namedtuple('FullArgSpec',
+                           ('args', 'varargs', 'varkw', 'defaults',
+                            'kwonlyargs', 'kwonlydefaults', 'annotations'))
+        return ret_t(args, varargs, keywords, defaults, [], None, {})
+
+
+def parse_function(func, input_names=None):
+    """Collect Function-relevant information from a Python function
+
+    Parameters
+    ----------
+    func : callable, source code of a function, or Undefined
+
+    Returns
+    -------
+    arg_names : list of str
+        Names of all named arguments
+    default_args : dict of (str, object) pairs
+        Names and default values of arguments with default values
+    kwargs : {True, False}
+        Name of keyword argument dictionary, if given
+    """
+    if isinstance(func, (str, bytes)):
+        function_str = func
+        func = create_function_from_source(func)
+    elif hasattr(func, '__call__'):
+        function_str = getsource(func)
+    else:
+        raise TypeError('Unknown type of function')
+
+    argspec = getfullargspec(func)
+    all_args = argspec.args + argspec.kwonlyargs
+    # Arguments with defaults must occur at the end of the protocol
+    all_defaults = ({} if argspec.defaults is None else
+                    dict(zip(argspec.args[-len(argspec.defaults):],
+                             argspec.defaults)))
+    if argspec.kwonlydefaults is not None:
+        all_defaults.update(argspec.kwonlydefaults)
+
+    required_args = set(all_args) - set(all_defaults.keys())
+    if input_names is None:
+        input_names = all_args
+    else:
+        missed_args = required_args - set(input_names)
+        if missed_args:
+            raise ValueError('These positional arguments must be in '
+                             'input names: {}'.format(', '.join(missed_args)))
+
+        unknown = set(input_names) - set(all_args)
+        if unknown:
+            raise ValueError('Input names do not match function arguments: '
+                             '{}'.format(', '.join(unknown)))
+
+    banned_names = list(set(all_defaults.keys()) - set(input_names))
+
+    return (function_str, all_args, all_defaults, banned_names,
+            argspec.varkw is not None)
+
+
+class FunctionInputSpec(BaseInterfaceInputSpec):
+    function_str = traits.Str(mandatory=True,
+                              desc='code for function')
+
+    def __deepcopy__(self, memo):
+        from copy import deepcopy
+        values = deepcopy(self.trait_get(), memo)
+        dup = self.clone_traits(memo=memo)
+        orig_traits = self.traits()
+        dup_traits = dup.traits()
+        for name, value in values.items():
+            if name not in dup_traits:
+                dup.add_trait(name, deepcopy(orig_traits[name], memo))
+        dup.trait_set(**values)
+        return dup
```

Comment:
Okay, it's the input spec. There's something about moving to a (less) dynamic input spec that breaks the following tests:

* `nipype/pipeline/plugins/tests/test_multiproc_nondaemon.py` (all)
* `nipype/pipeline/engine/tests/test_engine.py::test_serial_input
* `nipype/pipeline/engine/tests/test_workflows.py::test_outputs_removal_wf[MultiProc-*-*]`

It's not throwing errors, and I'm having trouble tracking down the specific code that is deadlocking.